File: haskell/lib/Kan/Invar_day.hs
module Invar_day where

import Control.Arrow ((&&&))

data Day f g a = Day (f b) (g c) (b -> c -> a) (a -> (b, c))

day :: f a -> g b -> Day f g (a, b)
day fa gb = Day fa gb (,) (\(a, b) -> (a, b))

invmap :: (a -> b) -> (c -> d) -> Day f g a -> Day f g b
invmap f g (Day fb gc bca abc) = Day fb gc (f *** g >>> bca) (abc >>> (\a -> let (b, c) = (g &&& id) a in (b, c)))

assoc :: Day f (Day g h c) a -> Day (Day f g c) h a
assoc (Day fb (Day gd he dec cde) bca abc) =
  let f a = let (b, c) = abc a in let (d, e) = cde c in ((b, d), e)
      g (b, d) e = bca b (dec d e)
  in Day (Day fb gd (id &&& fst) snd) he g f

disassoc :: Day (Day f g b) h a -> Day f (Day g h a) b
disassoc (Day (Day fb gc deb bde) hd bca abc) =
  let f e (d, c) = bca (deb e d) c
      g a = let (b, c) = abc a in let (d, e) = bde b in (d, (e, c))
  in Day fb (Day gc hd snd (id &&& fst)) f g

swapped :: Day f g a -> Day g f a
swapped (Day fb gc bca abc) = Day gc fb (bca *** id) (abc >>> (\a -> let (b, c) = a in (c, b)))

intro1 :: f a -> Day () f a
intro1 fa = Day () fa (const id) (id &&& snd)

intro2 :: f a -> Day f () a
intro2 fa = Day fa () (fst &&& const ())

elim1 :: Day () f b -> f b
elim1 (Day () fc bca abc) = bca () <$> fc

elim2 :: Day f () b -> f b
elim2 (Day fb () bca abc) = flip bca () <$> fb

trans1 :: (f a -> g b) -> Day f h a -> Day g h a
trans1 fg (Day fb hc bca abc) = Day (fg fb) hc bca abc

trans2 :: (g a -> h b) -> Day f g a -> Day f h a
trans2 gh (Day fb gc bca abc) = Day fb (gh gc) bca abc

toContravariant :: Day f g a -> Contravariant.Day f g a
toContravariant (Day fb gc _ abc) = Contravariant.Day fb gc abc

toCovariant :: Day f g a -> Covariant.Day f g a
toCovariant (Day fb gc bca _) = Covariant.Day fb gc bca

File: ocaml/lib/kan/invar_day.ml
module Day = struct
  type ('f, 'g, 'a) t = forall 'b 'c. 'f 'b * 'g 'c * ('b -> 'c -> 'a) * ('a -> ('b * 'c))

  let day (fa : 'a 'f) (gb : 'b 'g) : ('a, 'b, 'a * 'b) t =
    (fa, gb, (fun a b -> (a, b)), (fun (a, b) -> (a, b)))

  let invmap (f : 'a -> 'b) (g : 'c -> 'd) (Day (fb, gc, bca, abc)) : ('f, 'g, 'b) t =
    Day (fb, gc, (fun b c -> f (bca b c)), (fun a -> let (b, c) = abc (g a) in (b, c)))

  let assoc (Day (fb, Day (gd, he, dec, cde), bca, abc)) : (('f, 'g, 'c) t, 'h, 'a) t =
    let f a =
      let (b, c) = abc a in
      let (d, e) = cde c in
      ((b, d), e)
    and g (b, d) e = bca b (dec d e) in
    Day (Day (fb, gd, (fun (b, d) -> b), (fun c -> c)), he, g, f)

  let disassoc (Day (Day (fb, gc, deb, bde), hd, bca, abc)) : ('f, ('g, 'h, 'a) t, 'a) t =
    let f e (d, c) = bca (deb e d) c
    and g a =
      let (b, c) = abc a in
      let (d, e) = bde b in
      (d, (e, c)) in
    Day (fb, Day (gc, hd, (fun (c, e) -> c), (fun a -> a)), f, g)

  let swapped (Day (fb, gc, bca, abc)) : ('g, 'f, 'a) t =
    Day (gc, fb, (fun c b -> bca b c), (fun a -> let (b, c) = abc a in (c, b)))

  let intro1 (fa : 'a 'f) : (unit, 'f, 'a) t = Day (()), fa, (fun () a -> a), (fun a -> ((), a)))
  let intro2 (fa : 'a 'f) : ('f, unit, 'a) t = Day (fa, (), (fun a () -> a), (fun a -> (a, ())))
  let elim1 (Day (()), fc, bca, abc)) : 'c 'f = invmap (bca ()) (snd <@ abc) fc
  let elim2 (Day (fb, (), bca, abc)) : 'b 'f = invmap (flip bca ()) (fst <@ abc) fb

  let trans1 (fg : 'a 'f -> 'b 'g) (Day (fb, hc, bca, abc)) : ('g, 'h, 'a) t =
    Day (fg fb, hc, bca, abc)

  let trans2 (gh : 'a 'g -> 'b 'h) (Day (fb, gc, bca, abc)) : ('f, 'h, 'a) t =
    Day (fb, gh gc, bca, abc)

  let toContravariant (Day (fb, gc, _, abc)) = Contravariant.Day (fb, gc, abc)
  let toCovariant (Day (fb, gc, bca, _)) = Covariant.Day (fb, gc, bca)
end